Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp

```diff
@@ -1128,6 +1129,69 @@ void MediaPlayerPrivateGStreamer::notifyPlayerOfVideoCaps()
     m_player->mediaEngineUpdated();
 }
 
+String audioTrackIdFromStreamStart(GstEvent* streamStart) {
+    const gchar* streamIdAsCharacters;
+    gst_event_parse_stream_start(streamStart, &streamIdAsCharacters);
+
+    String streamId;
+    if (streamIdAsCharacters) {
+        StringView streamIdView(streamIdAsCharacters);
+        size_t position = streamIdView.find('/');
+
+        if (position != notFound && position + 1 < streamIdView.length())
+            streamId = streamIdView.substring(position + 1).toString();
+    }
+
+    return streamId;
+}
+
+class StreamStartProbeData : public ThreadSafeRefCounted<StreamStartProbeData> {
+public:
+    StreamStartProbeData(MediaPlayerPrivateGStreamer* player, int trackIndex)
+        : player(player)
+        , trackIndex(trackIndex)
+    { }
+
+    MediaPlayerPrivateGStreamer* player;
+    int trackIndex;
+    bool probeWasRemoved { false };
+    Lock lock;
+};
+
+GstPadProbeReturn MediaPlayerPrivateGStreamer::streamStartProbeCallback(GstPad*, GstPadProbeInfo* info, gpointer data)
+{
+    Ref<StreamStartProbeData> probeData(*static_cast<StreamStartProbeData*>(data));
+    LockHolder locker(probeData->lock);
+    GstEvent* event = GST_PAD_PROBE_INFO_EVENT(info);
+
+    if (UNLIKELY(probeData->probeWasRemoved) || GST_EVENT_TYPE(event) != GST_EVENT_STREAM_START)
+        return GST_PAD_PROBE_OK;
+
+    String streamId = audioTrackIdFromStreamStart(event);
+    bool ok = false;
+    auto audioTracks = probeData->player->m_audioTracks;
+    auto mediaPlayer = probeData->player->m_player;
+
+    for (auto& track : audioTracks.values()) {
+        if (track->trackIndex() == probeData->trackIndex) {
+            mediaPlayer->removeAudioTrack(*track);
+            track->setId(streamId);
+            mediaPlayer->addAudioTrack(*track);
+            mediaPlayer->mediaEngineUpdated();
```

Comment:
I'd probably leave the update out of the loop and do it outside only if ok is true.

Reply:
@calvaris, as this code is part of the probe, it can potentially (and likely) run in a non-main thread. I'm not sure it's a good idea to call MediaPlayer (ie: WebCore) code from a non-main thread. What do you think about that?

Reply:
I hadn't checked that, actually. I don't know if the code is protected or if it needs protection. If it is, you can maybe defer that to the main thread.